server/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java

```diff
@@ -0,0 +1,236 @@
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              try {
+                runnable.run();
+                postRunnable.run();
+              }catch(RuntimeException e){
+                // Don't leave stale cache on error
+                delete(cacheId);
+                throw e;
+              }
+            }
+          },
+          updateMs - System.currentTimeMillis() % updateMs, updateMs, TimeUnit.MILLISECONDS
+      );
+    } else {
+      future = listeningScheduledExecutorService.schedule(runnable, 0, TimeUnit.MILLISECONDS);
+      Futures.addCallback(
+          future, new FutureCallback<Object>()
+          {
+            @Override
+            public void onSuccess(Object result)
+            {
+              try {
+                postRunnable.run();
+              }catch(RuntimeException e){
+                delete(cacheId);
+                throw e;
+              }
+            }
+
+            @Override
+            public void onFailure(Throwable t)
+            {
+              // Don't leave stale cache on error
+              delete(cacheId);
+              log.error(t, "Error populating namespace [%s]", namespace.toString());
+            }
+          }
+      );
+    }
+    final ListenableFuture<?> otherFuture = futures.putIfAbsent(namespace.getNamespace(), future);
+    if(otherFuture != null){
+      future.cancel(true);
+      throw new IAE("Namespace [%s] already exists! Leaving prior running", namespace.toString());
+    }
+    return future;
+  }
+
+  protected abstract boolean swapAndClearCache(String namespaceKey, String cacheKey);
+  public abstract ConcurrentMap<String, String> getCacheMap(String namespaceKey);
+
+  public boolean delete(final String ns)
+  {
+    if(!fnCache.containsKey(ns)){
+      return false;
+    }
+    disabledNamespaces.add(ns);
+    final ListenableFuture<?> future = futures.get(ns);
+    future.cancel(false); // let it finish first.
+    final CountDownLatch latch = new CountDownLatch(1);
+    final AtomicBoolean done = new AtomicBoolean(false);
+    Futures.addCallback(
+        future, new FutureCallback<Object>()
+        {
+          @Override
+          public void onSuccess(Object result)
+          {
+            done.set(true);
+            latch.countDown();
+          }
+
+          @Override
+          public void onFailure(Throwable t)
+          {
+            // Expect CancellationException
+            done.set(true);
+            latch.countDown();
+            if(!(t instanceof CancellationException)){
+              log.error(t, "Error in parsing of namespace [%s]", ns);
+            }
+          }
+        }
+    );
+    try {
+      latch.await();
+    }
+    catch (InterruptedException e) {
+      throw Throwables.propagate(e);
+    }
+    // This is really just a memory barrier. I could have spinlocked on this but didn't want to burn CPU cycles
+    if(!done.get()){
```

Comment:
what was the error?
In any case, I don't think having "done" there is helping at all bcoz  "IF latch.await() returns without exception THEN done is set to true"